### PR TITLE
MINIFICPP-2022 Add valid repository size metrics for all repositories

### PR DESCRIPTION
**Pull request by lordgamez**
Calculate and report repository sizes (memory for volatile, disk space for persistent repositories) on C2 and Prometheus interfaces. Additionally add max repository size metric where available.

https://issues.apache.org/jira/browse/MINIFICPP-2022

Depends on https://github.com/apache/nifi-minifi-cpp/pull/1485 and https://github.com/apache/nifi-minifi-cpp/pull/1489

-------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: libminifi/include/core/Repository.h [hunk 1]
```diff
@@ -54,28 +54,22 @@ constexpr auto MAX_REPOSITORY_STORAGE_SIZE = 10_MiB;
 constexpr auto MAX_REPOSITORY_ENTRY_LIFE_TIME = std::chrono::minutes(10);
 constexpr auto REPOSITORY_PURGE_PERIOD = std::chrono::milliseconds(2500);
 
-class Repository : public core::CoreComponent {
+class Repository : public core::ReportableRepository {
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/core/Repository.h:57]: You are right, I also felt this should be the other way around, but the root cause of this problem is the current namings of the repositories. The `core::Repository` class should be the most generic class for all repositories, but it is only the base for the flowfile and provenance repositories. `ContentRepository` was entirely separate from the hierarchy starting with `core::Repository` so I had to come up with something new as I didn't want to change the current class names (core::Repository is used way too many times and I didn't want to blow up this PR). 

Maybe we can rename the `ReportableRepository` to `Repository` and come up with new names to the classes we already have. The main difference seems to be that `core::Repository` and its children are discrete key based storage repositories with put, multiput, delete, get operations while `ContentRepository` is stream based repositories with read, write operations. So the `Repository` could be renamed to `KeyValueRepository` which is inherited by `ThreadedKeyValueRepository`, but I would create a separate PR for that.

Do you think this is viable?

File: extensions/rocksdb-repos/database/OpenRocksDb.cpp [hunk 2]
```diff
@@ -118,8 +114,18 @@ rocksdb::DB* OpenRocksDb::get() {
...
-}  // namespace apache
-}  // namespace org
+std::optional<uint64_t> OpenRocksDb::getApproximateSizes() const {
+  rocksdb::SizeApproximationOptions options;
+  options.include_memtabtles = true;
+  std::array<rocksdb::Range, 1> ranges;
+  ranges[0].start = "";
+  ranges[0].limit = "~";
+  uint64_t value = 0;
+  auto status = impl_->GetApproximateSizes(options, column_->handle.get(), ranges.data(), 1, &value);
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply extensions/rocksdb-repos/database/OpenRocksDb.cpp]: Updated in 38e9c97ff6fbcfb6479eec3615c025d076d1274b. Yes `include_memtabtles` us a typo in the rocksdb API :)

File: libminifi/include/core/Repository.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/core/Repository.h]: Sounds good to me, updated in 546dcf6caab7270d3d010c123af2739857e4120c

[the remark this answers is not in the document]
Comment (szaszm) [reply libminifi/include/core/Repository.h]: Looks good. Forgot to mention the design principle behind my recommendation: [Interface Segregation Principle](https://en.wikipedia.org/wiki/Interface_segregation_principle) (I in SOLID)

Whoever depends on `RepositoryMetricsSource` does not need to depend on `CoreComponent`.

File: extensions/rocksdb-repos/ProvenanceRepository.cpp [hunk 3]
```diff
@@ -145,9 +112,12 @@ void ProvenanceRepository::destroy() {
...
 
 uint64_t ProvenanceRepository::getKeyCount() const {
+  auto opendb = db_->open();
+  if (!opendb) {
+    return 0;
+  }
   std::string key_count;
-  db_->GetProperty("rocksdb.estimate-num-keys", &key_count);
-
+  opendb->GetProperty("rocksdb.estimate-num-keys", &key_count);
```
Comment (adamdebreceni) [inline extensions/rocksdb-repos/ProvenanceRepository.cpp:120]: is this different from `getRepositoryEntryCount`?

File: extensions/rocksdb-repos/database/OpenRocksDb.cpp [hunk 4]
```diff
@@ -118,8 +114,15 @@ rocksdb::DB* OpenRocksDb::get() {
...
 }
 
-}  // namespace internal
-}  // namespace minifi
-}  // namespace nifi
-}  // namespace apache
-}  // namespace org
+std::optional<uint64_t> OpenRocksDb::getApproximateSizes() const {
+  const rocksdb::SizeApproximationOptions options{ .include_memtabtles = true };
+  const rocksdb::Range range("", "~");
```
Comment (adamdebreceni) [inline extensions/rocksdb-repos/database/OpenRocksDb.cpp:119]: could keys with leading `~` exist? content paths are prefixed with the content dir's path, I could not find if we turn that into absolute path or not, but if not, could the user set it relative to their home (e.g. `~/content_repository`)?

File: extensions/rocksdb-repos/ProvenanceRepository.cpp (hunk 3 above)
Comment (lordgamez) [reply extensions/rocksdb-repos/ProvenanceRepository.cpp]: Good point, as it was only used in tests, removed this in 2c1c6841b84cd569185323665d9d927a029396e9

File: extensions/rocksdb-repos/database/OpenRocksDb.cpp (hunk 4 above)
Comment (lordgamez) [reply extensions/rocksdb-repos/database/OpenRocksDb.cpp]: Good point, I suppose it's possible to have a path like this, updated in 2c1c6841b84cd569185323665d9d927a029396e9

File: libminifi/src/core/repository/VolatileRepository.cpp [hunk 5]
```diff
@@ -71,6 +71,9 @@ bool VolatileRepository::Put(const std::string& key, const uint8_t *buf, size_t
     }
   } while (!updated);
   repo_data_.current_size += size;
+  if (repo_data_.current_entry_count < repo_data_.max_count) {
+    ++repo_data_.current_entry_count;
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/src/core/repository/VolatileRepository.cpp:75]: If we store the `max_count` number of entries even if a new entry is added it overwrites the old entry in a round robin fashion, so if we reach the maximum number of entries we should not increase the value. The size seems to be reduced with the reclaimed size beforehand, but it's also only an estimate.

File: extensions/rocksdb-repos/FlowFileRepository.h [hunk 6]
```diff
@@ -108,25 +99,17 @@ class FlowFileRepository : public ThreadedRepository, public SwapManager {
...
-
-  std::thread& getThread() override {
-    return thread_;
-  }
-
+  std::filesystem::path checkpoint_dir_;
   moodycamel::ConcurrentQueue<std::string> keys_to_delete;
   std::shared_ptr<core::ContentRepository> content_repo_;
-  std::unique_ptr<minifi::internal::RocksDatabase> db_;
+  std::unique_ptr<rocksdb::Checkpoint> checkpoint_;
```
Comment (adamdebreceni) [inline extensions/rocksdb-repos/FlowFileRepository.h:110]: was this added by accident? (and the `checkpoint_dir_` member)

Comment (lordgamez) [reply extensions/rocksdb-repos/FlowFileRepository.h]: Good catch, had some troubles while rebasing, I left it there by accident, updated in 5affb58bde8cb1fd4ec2424f60064a53a48bf12c

**Review by adamdebreceni — APPROVED** [review]

File: libminifi/test/unit/MetricsTests.cpp [hunk 7]
```diff
@@ -126,6 +126,15 @@ TEST_CASE("RepositorymetricsHaveRepo", "[c2m4]") {
 
     REQUIRE("size" == size.name);
     REQUIRE("0" == size.value);
+
+    minifi::state::response::SerializedResponseNode max_size = resp.children.at(3);
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/test/unit/MetricsTests.cpp]: Good idea, updated in 5ed288e89b50d48b0e966a60ede12fd25377fe0f

File: libminifi/src/core/state/nodes/ResponseNodeLoader.cpp [hunk 8]
```diff
@@ -32,11 +32,12 @@
 
 namespace org::apache::nifi::minifi::state::response {
 
-ResponseNodeLoader::ResponseNodeLoader(std::shared_ptr<Configure> configuration, std::shared_ptr<core::Repository> provenance_repo,
-    std::shared_ptr<core::Repository> flow_file_repo, std::shared_ptr<core::FlowConfiguration> flow_configuration)
+ResponseNodeLoader::ResponseNodeLoader(std::shared_ptr<Configure> configuration, std::shared_ptr<core::RepositoryMetricsSource> provenance_repo,
+      std::shared_ptr<core::RepositoryMetricsSource> flow_file_repo, std::shared_ptr<core::RepositoryMetricsSource> content_repo, std::shared_ptr<core::FlowConfiguration> flow_configuration)
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/src/core/state/nodes/ResponseNodeLoader.cpp]: Good idea, updated in 3654c584f3e1399b46df0560d536d62786ca2ce3

File: libminifi/test/rocksdb-tests/RepoTests.cpp [hunk 9]
```diff
@@ -515,4 +519,221 @@ TEST_CASE("FlowFileRepository synchronously pushes existing flow files") {
...
+  repository->initialize(configuration);
+  repository->loadComponent(content_repo);
+  auto original_content_repo_size = content_repo->getRepositorySize();
+
+  auto flow_file = std::make_shared<minifi::FlowFileRecord>();
+
+  auto content_session = content_repo->createSession();
+  auto claim = content_session->create();
+  auto stream = content_session->write(claim);
+  stream->write(reinterpret_cast<const uint8_t*>(content.data()), content.size());
```
Comment (szaszm) [inline libminifi/test/rocksdb-tests/RepoTests.cpp:654]:
```suggestion
  stream->write(gsl::make_span(content).as_span<const std::byte>());
```

Comment (lordgamez) [reply libminifi/test/rocksdb-tests/RepoTests.cpp]: Updated in d192f3dadee8367c0b10084a228f36ca5bf6d457

File: METRICS.md [hunk 10]
```diff
@@ -100,17 +100,19 @@ QueueMetrics is a system level metric that reports queue metrics for every conne
...
+| repository_size_bytes     | repository_name | Current size of the repository                  |
+| max_repository_size_bytes | repository_name | Maximum size of the repository (0 if unlimited) |
+| repository_entry_count    | repository_name | Current number of entries in the repository     |
 
-| Label                    | Description                                                     |
-|--------------------------|-----------------------------------------------------------------|
-| repository_name          | Name of the reported repository                                 |
+| Label                    | Description                                                                                                                           |
+|--------------------------|---------------------------------------------------------------------------------------------------------------------------------------|
+| repository_name          | Name of the reported repository. There are three repositories present with the following names: `flowile`, `content` and `provenance` |
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply METRICS.md]: Updated in 4c763c831909cd5dc9cfa8116e6a5599eb68193e

File: libminifi/include/core/repository/FileSystemRepository.h [hunk 11]
```diff
@@ -50,6 +52,18 @@ class FileSystemRepository : public core::ContentRepository {
 
   void clearOrphans() override;
 
+  uint64_t getRepositorySize() const override {
+    return utils::file::FileUtils::path_size(directory_);
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/core/repository/FileSystemRepository.h]: Updated in 4c763c831909cd5dc9cfa8116e6a5599eb68193e

File: libminifi/include/core/repository/FileSystemRepository.h [hunk 12]
```diff
@@ -50,6 +52,18 @@ class FileSystemRepository : public core::ContentRepository {
...
+  }
+
+  uint64_t getRepositoryEntryCount() const override {
+    auto dir_it = std::filesystem::directory_iterator(directory_);
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply libminifi/include/core/repository/FileSystemRepository.h]: Updated in 4c763c831909cd5dc9cfa8116e6a5599eb68193e

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**